### PR TITLE
Add SiteExtension.proj to replace build-extension.ps1

### PR DESCRIPTION
**Pull request by jviau**
### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [ ] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR -- TODO
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

This PR introduces a new project `WebHost.Script.SiteExtension`, which is a replacement for `build-extension.ps1`. This project can be used via `dotnet` CLI to restore, build, and publish all the artifacts needed for the Antares site extension. Note that this project has _support_ for multi-targeting, but only targets a single TFM. This is in preparation for porting to `in-proc` branch to keep the diff minimal between the two.

### Other Changes

- Removes `update-hostreferences.ps1`, this is no longer used anywhere
- Groups all worker versions into `workers.props` file. Left python out due to code owners, but ideally python will be moved there as well.
- zip-after-publish is extracted into a set of re-usable targets `ZipPublish.targets`.
- `RuntimeIdentifiers` added to `WebHost.csproj`. Having this set of identifiers listed improves the restore task (otherwise each time you want to use a different runtime, you have to re-restore).

### CODE REVIEW

File: eng/ci/templates/official/jobs/build-artifacts-windows.yml
Comment (kshyju) [inline eng/ci/templates/official/jobs/build-artifacts-windows.yml]: Shall we remove the "release/inproc6/4" and "release/inproc8/4" usage here?

File: test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj
Comment (kshyju) [inline test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj]: The tests are not using anything from these packages?

File: src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj
Comment (kshyju) [inline src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj]: Are we skipping [this](https://github.com/Azure/azure-functions-host/blob/c6365fdbba0f895b24da785a9b2f31ad0a759530/build/build-extensions.ps1#L103)?

File: eng/ci/templates/official/jobs/build-artifacts-windows.yml
Comment (jviau) [reply eng/ci/templates/official/jobs/build-artifacts-windows.yml]: I have updated that in a different PR: #10171

File: src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj
Comment (jviau) [reply src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj]: See worker.props `CleanDotnetWorkerFiles` target

File: test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj
Comment (jviau) [reply test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj]: They are now brought in through `worker.props` at the top.

**Review by kshyju — APPROVED** [review]
Approving to unblock.

**Review by liliankasem — APPROVED** [review]

File: build/workers.props
Comment (fabiocav) [inline build/workers.props]: Might be overkill, but this makes me think about having a file for each worker, all imported by this one. We could keep all of the worker specific logic there and code ownership would also be more granular.

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj]: Left a comment above about worker specific files. Would help break worker specific logic like this up.

File: src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj
Comment (fabiocav) [inline src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj]: Could this move to `python.props` conditional on the RID? Ultimately, Python is excluded from Windows builds of the host.

**Review by fabiocav — APPROVED** [review]
Left a few comments. Reviewed best I could based on the nature of the changes. Can we please make sure this is rebased on latest and compare the artifact output to ensure we're good (assuming this was already done, but worth calling out).

File: src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj
Comment (jviau) [reply src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj]: @fabiocav I was under the impression it would be needed for local development of windows. I wanted it still included in the `WebHost.csproj`.

Comment (fabiocav) [reply src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj]: Unless there's a scenario I'm missing, we should be OK as those flows won't specify a RID. We could use a different condition check if needed, but don't treat this as a blocker. Just thought it would be nice to have it all in that same file.

File: build/workers.props
Comment (jviau) [reply build/workers.props]: Done

File: src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj
Comment (jviau) [reply src/WebJobs.Script.SiteExtension/WebJobs.Script.SiteExtension.proj]: Done. It _works_, but a bit hacky. For `SiteExtension.proj`, works fine as RID is set for both restore & build. For `WebHost.csproj`, RID is set only for build, so restore does incorporate the python worker targets still. But they then get removed from the `SuppressPythonWorker` target

File: eng/build/Workers.Python.props
Comment (gavin-aguiar) [inline eng/build/Workers.Python.props:9]: @jviau I didn't completely understand this step? 
Is this step required because this [PR](https://github.com/Azure/azure-functions-python-worker/pull/1472) is not merged yet?

**Review by gavin-aguiar — APPROVED** [review]

File: eng/build/Workers.Python.props
Comment (hallvictoria) [reply eng/build/Workers.Python.props:9]: Following up on this, it looks like the worker is only included if the runtime isn't windows. The step below removes the worker for windows, though. 

Shouldn't the worker already not be included if the runtime is windows? Why would the removal step be necessary?

Comment (jviau) [reply eng/build/Workers.Python.props:9]: We have not been including the python worker for windows site extension for a while. I believe it is already on the machine (@fabiocav to clarify).

Today this is done as part of the `build-extensions.ps1` script. It is removed after msbuild copies it. This approach avoids the copy altogether when building windows runtimes.
https://github.com/Azure/azure-functions-host/blob/8ba99f7b80e2e7ffbfe32fe6fdc15efb61985d55/build/build-extensions.ps1#L95

Note: this does not affect _local_ development, where `RuntimeIdentifier=""`, so python will be copied then.

@gavin-aguiar its tangential to that PR you linked. That PR is to refactor python targets to make it play nicer with other MSBuild targets (avoid using such general names like `Initialize`, `SourceFiles`, as well as participate in an existing copy task)

**Comment by kamperiadis** [comment]
@jviau Just as a heads up, we have recently updated the java worker version so that appears to be causing conflicts for your PR.

**Comment by liliankasem** [comment]
@jviau is this PR still relevant?

**Comment by jviau** [comment]
@liliankasem yes, I plan on merging it in soon. Was just waiting for the current release to be cut. But that has been done so I will update this PR soon :)

**Pull request merged**
